src/orion/client/runner.py

```diff
@@ -0,0 +1,334 @@
+
+                if self.has_remaining and not self.is_done:
+                    msg = (
+                        f"{msg}; worker has leg room (has_remaining: {self.has_remaining})"
+                        f" and optimization is not done (is_done: {self.is_done})"
+                    )
+
+                raise LazyWorkers(msg)
+
+        return self.trials
+
+    def should_sample(self):
+        """Check if more trials could be generated"""
+
+        if self.is_broken or self.is_done:
+            return 0
+
+        pending = len(self.pending_trials) + self.trials
+        remains = self.max_trials_per_worker - pending
+
+        n_trial = min(self.free_worker, remains)
+        should_sample_more = self.free_worker > 0 and remains > 0
+
+        return int(should_sample_more) * n_trial
+
+    def sample(self):
+        """Sample new trials for all free workers"""
+        n_trial = self.should_sample()
+
+        if n_trial > 0:
+            # the producer does the job of limiting the number of new trials
+            # already no need to worry about it
+            # NB: suggest reserve the trial already
+            new_trials = self._suggest_trials(n_trial)
+            log.debug(f"Sampled {len(new_trials)} new configs")
+            return new_trials
+
+        return []
+
+    def scatter(self, new_trials):
+        """Schedule new trials to be computed"""
+        new_futures = []
+        for trial in new_trials:
+            future = self.client.executor.submit(
+                _optimize, trial, self.fct, self.trial_arg, **self.kwargs
+            )
+            self.pending_trials[future] = trial
+            new_futures.append(future)
+
+        self.futures.extend(new_futures)
+        log.debug("Scheduled new trials")
+
+    def gather(self):
+        """Gather the results from each worker asynchronously"""
+        results = []
+        try:
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except KeyboardInterrupt:
+            self._release_all()
+            raise
+
+        to_be_raised = None
+        log.debug(f"Gathered new results {len(results)}")
+
+        # register the results
+        for result in results:
+            trial = self.pending_trials.pop(result.future)
+
+            if isinstance(result, AsyncResult):
+                # NB: observe release the trial already
+                self.client.observe(trial, result.value)
+                self.trials += 1
+
+            if isinstance(result, AsyncException):
+                if (
+                    isinstance(result.exception, ExecutionError)
+                    and result.exception.return_code == self.interrupt_signal_code
+                ):
+                    to_be_raised = KeyboardInterrupt()
+                    self.client.release(trial, status="interrupted")
+                    continue
+
+                if isinstance(result.exception, InvalidResult):
```

Comment:
Follow up from this [comment](https://github.com/Epistimio/orion/pull/684#discussion_r762087196) here. InvalidResult is raised when the experiment tries to observe a trial. We should not expect it to occur during trial execution, but it could occur during self.client.observe a few lines above. That's where it should be handled.

Reply:
move the catching of invalid result

Reply:
The trial should also be released and set to broken. This would also deserve a specific test.

Reply:
added the test & the broken release

Reply:
Thanks!